shards/aggregate.go

```diff
@@ -157,30 +151,15 @@ func newFlushCollectSender(opts *zoekt.SearchOptions, sender zoekt.Sender) (zoek
 	}), finalFlush
 }
 
-// limitSender wraps a sender and calls cancel once it has seen either of
-// `docLimit` files or `matchLimit` matches.
-func limitSender(cancel context.CancelFunc, sender zoekt.Sender, docLimit, matchLimit int, chunkMatches bool) zoekt.Sender {
-	docLimited := docLimit > 0
-	matchLimited := matchLimit > 0
+// limitSender wraps a sender and calls cancel once the truncator has finished
+// truncating.
+func limitSender(cancel context.CancelFunc, sender zoekt.Sender, truncator zoekt.DisplayTruncator) zoekt.Sender {
 	return stream.SenderFunc(func(result *zoekt.SearchResult) {
-		// We cancel when we hit either of the two limits. It is assumed that,
-		// after canceling, we are not called again.
-
-		if docLimited {
-			if len(result.Files) >= docLimit {
-				result.Files = result.Files[:docLimit]
-				cancel()
-			}
-			docLimit -= len(result.Files)
+		var hasMore bool
+		result.Files, hasMore = truncator(result.Files)
```

Comment:
a `stream.SenderFunc` can be called concurrently, right? The current `truncator` doesn't look like it's safe to call concurrently

Reply:
Agree, I don't see how you've resolved the race condition you pointed out on my PR.

Reply:
I think you have found a bug! I took a quick look and if you don't have FlushWallTime set I think there is indeed a race condition. I will tackle that tomorrow, since this issue existed before this PR.

Reply:
https://github.com/sourcegraph/zoekt/issues/631